ouroboros-consensus-test/src/Test/Util/ChainUpdates.hs

```diff
@@ -0,0 +1,184 @@
+
+genChainUpdates' ::
+     UpdateBehavior
+  -> SecurityParam
+  -> Int
+  -> ChainUpdateState
+  -> Gen ChainUpdateState
+genChainUpdates' updateBehavior securityParam n =
+    execStateT (replicateM_ n genChainUpdate)
+  where
+    -- Modify the state
+    addUpdate u cus = cus { cusUpdates = u : cusUpdates cus }
+    setChain  c cus = cus { cusCurrentChain = c }
+
+    k = fromIntegral $ maxRollbacks securityParam
+
+    genChainUpdate = do
+      ChainUpdateState { cusCurrentChain = chain } <- get
+      let genValid =
+            frequency'
+              [ (3, genAddBlock Valid)
+              , ( if Chain.null chain then 0 else 1
+                , genSwitchFork (choose (1, k))
+                )
+              ]
+      frequency' $
+        (5, replicateM_ 2 genValid) :
+        [ (1, genInvalidBlock) | updateBehavior == TentativeChainBehavior ]
+
+    genBlockToAdd validity = do
+        ChainUpdateState { cusCurrentChain = chain } <- get
+        block <- lift $ case Chain.head chain of
+          Nothing      -> setValidity . firstBlock <$> genForkNo
+          Just curHead -> do
+            forkNo <- case validity of
+              Valid   ->  genForkNo
+              Invalid -> pure 3
+            return
+              . modifyFork (const forkNo)
+              . setValidity
+              $ successorBlock curHead
+        modify $ setChain (Chain.addBlock block chain)
+        return block
+      where
+        setValidity b = b { tbValid = validity }
+        genForkNo = case validity of
+          Valid -> frequency
+            [ (1, return 0)
+            , (1, choose (1, 2))
+            ]
+          -- Blocks with equal hashes have to have equal validity, so we reserve
+          -- a specific ForkNo for invalid blocks to ensure this.
+          Invalid -> pure 3
+
+    genAddBlock validity = do
+      block <- genBlockToAdd validity
+      modify $ addUpdate (AddBlock block)
+
+    genSwitchFork genRollBackBlocks = do
+      ChainUpdateState { cusCurrentChain = chain } <- get
+      rollBackBlocks <- lift genRollBackBlocks
+      let chain' = Chain.drop rollBackBlocks chain
+      modify $ setChain chain'
+      blocks <- replicateM rollBackBlocks (genBlockToAdd Valid)
+      modify $ addUpdate (SwitchFork (Chain.headPoint chain') blocks)
+
+    genInvalidBlock = do
+      genAddBlock Invalid
+      genSwitchFork (pure 1)
+
+-- | Variant of 'frequency' that allows for transformers of 'Gen'
+frequency' :: (MonadTrans t, Monad (t Gen)) => [(Int, t Gen a)] -> t Gen a
+frequency' [] = error "frequency' used with empty list"
+frequency' xs0 = lift (choose (1, tot)) >>= (`pick` xs0)
+  where
+    tot = sum (map fst xs0)
+
+    pick n ((k,x):xs)
+      | n <= k    = x
+      | otherwise = pick (n-k) xs
+    pick _ _  = error "pick used with empty list"
+
+-- | Test that applying the generated updates gives us the same chain as
+-- '_currentChain'.
+prop_genChainUpdates :: SecurityParam -> Int -> Property
```

Comment:
It seemed like the easiest option to put this here (as this uses `ChainUpdateState` and related functions, which are not exported); but we could also create a `Test.Util.ChainUpdates.Internal` module as an alternative.